libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/fasterroute/RecordRouteObserverResults.kt

```diff
@@ -0,0 +1,100 @@
+package com.mapbox.navigation.core.internal.fasterroute
+
+import com.google.gson.GsonBuilder
+import com.google.gson.TypeAdapter
+import com.google.gson.stream.JsonReader
+import com.google.gson.stream.JsonWriter
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.directions.session.RoutesObserver
+import com.mapbox.navigation.core.directions.session.RoutesUpdatedResult
+import com.mapbox.navigation.core.routealternatives.AlternativeRouteMetadata
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import java.io.File
+
+/***
+ * Records routes updates to record-routes-observer folder.
+ * You can copy records from device file system to unit test resources and replay them using
+ * com.mapbox.navigation.core.fasterroute.ReplayRouteObserverResultsKt.readRouteObserverResults
+ */
+class RecordRouteObserverResults(
+    private val navigation: () -> MapboxNavigation
+) : RoutesObserver {
+
+    private var scope = CoroutineScope(Dispatchers.IO)
+    private var routesChangedCounter: Int = 0
+    private lateinit var preparation: Deferred<File>
+
+    override fun onRoutesChanged(result: RoutesUpdatedResult) {
+        if (routesChangedCounter == 0) {
+            prepareFolders()
+        }
+        val content = result.navigationRoutes
+            .map { route ->
+                listOf(
+                    route.routeIndex.toString(),
+                    route.routeOptions.toUrl("***").toString(),
+                    route.directionsResponse.toJson(),
+                    navigation().getAlternativeMetadataFor(route)?.let { toJson(it) }.orEmpty()
+                )
+            }
+            .flatten()
+            .joinToString(separator = "\n") { it }
+        val alternativesIds = result.navigationRoutes.drop(1)
+            .map { navigation().getAlternativeMetadataFor(it)!!.alternativeId }
+            .joinToString(separator = ",") { it.toString() }
+        writeOnDisk("$routesChangedCounter-${result.reason}-$alternativesIds.txt", content)
+        routesChangedCounter++
+    }
+
+    private fun toJson(metadata: AlternativeRouteMetadata): String {
+        val gson = GsonBuilder()
+            .registerTypeAdapter(
+                NavigationRoute::class.java,
```

Comment:
yep. gson can serialise `AlternativeRouteMetadata` on its own using reflections. But I don't want it to serialize whole navigation route + I doubt that it can this specific field. So I customise only navigation route serialisation